### PR TITLE
Sync V8 heap object type hierarchy with upstream

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: src/object.rs [hunk 1]
```diff
@@ -379,11 +380,14 @@ impl Object {
...
   }
 
-  /// The `Object` specific equivalent of `Data::get_hash()`.
-  /// This function is kept around for testing purposes only.
+  /// Returns the V8 hash value for this value. The current implementation
+  /// uses a hidden property to store the identity hash.
+  ///
+  /// The return value will never be 0. Also, it is not guaranteed to be
+  /// unique.
   #[doc(hidden)]
```
Comment (bartlomieju) [inline src/object.rs:388]: Was removal of this attribute omitted on purpose?

File: src/binding.cc [hunk 2]
```diff
@@ -2581,7 +2625,7 @@ v8::Isolate* v8__internal__GetIsolateFromHeapObject(const v8::Data& data) {
              : nullptr;
 }
 
-int v8__internal__Object__GetHash(const v8::Data& data) {
+int v8__internal__Object__GetHash(const v8::Value& data) {
```
Comment (bartlomieju) [inline src/binding.cc:2628]: A bit confusing, function name says `Object` but it's `v8::Value` passed. Can you add a comment here?

File: src/data.rs [hunk 3]
```diff
@@ -92,6 +140,10 @@ macro_rules! impl_try_from {
     impl<'s> TryFrom<Local<'s, $source>> for Local<'s, $target> {
       type Error = DataError;
       fn try_from(l: Local<'s, $source>) -> Result<Self, Self::Error> {
+        #[allow(dead_code)]
+        fn cast<T>(l: Local<$source>) -> Local<T> {
+          unsafe { transmute(l) }
+        }
```
Comment (bartlomieju) [inline src/data.rs:146]: Remove?

File: src/object.rs (hunk 1 above)
Comment (piscisaureus) [reply src/object.rs]: No, it was omitted by accident, thanks for pointing that out.

File: src/data.rs (hunk 3 above)
Comment (piscisaureus) [reply src/data.rs:147]: It's not actually dead code. I've added a comment.

File: src/binding.cc (hunk 2 above)
Comment (piscisaureus) [reply src/binding.cc]: I've renamed the function to `v8__Value__GetHash`.
PS: it's true that that V8 uses different class names than they do in the external API.
`i::Object` corresponds most closely to `Data`, but this particular function works only for true javascript values.

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks for addressing the comments

**Pull request merged**